server/src/main/java/org/elasticsearch/action/support/single/instance/TransportInstanceSingleOperationAction.java

```diff
@@ -257,26 +257,8 @@ public void onTimeout(TimeValue timeout) {
         }
     }
 
-    private class ShardTransportHandler implements TransportRequestHandler<Request> {
-
-        @Override
-        public void messageReceived(final Request request, final TransportChannel channel, Task task) throws Exception {
-            threadPool.executor(executor(request.shardId)).execute(new AbstractRunnable() {
-                @Override
-                public void onFailure(Exception e) {
-                    try {
-                        channel.sendResponse(e);
-                    } catch (Exception inner) {
-                        inner.addSuppressed(e);
-                        logger.warn("failed to send response for " + shardActionName, inner);
-                    }
-                }
-
-                @Override
-                protected void doRun() {
-                    shardOperation(request, ActionListener.wrap(channel::sendResponse, this::onFailure));
-                }
-            });
-        }
```

Comment:
ChannelActionListener also increments the ref counter:
https://github.com/elastic/elasticsearch/blob/f1185d6c4717ba361f0820969dc3667c409114e7/server/src/main/java/org/elasticsearch/action/support/ChannelActionListener.java#L31
and expects `channel::sendResponse` to decrement it back.

So existing approach must result in ref count goes < 0. I am surprised nothing surfaced it

Reply:
Today this is only used with responses that do not have any meaningful refcounting (i.e. they behave like `org.elasticsearch.core.RefCounted#ALWAYS_REFERENCED`). I added 11382abc2e5cad32b8d62b4cb1cb3ca69f92f287 anyway for the sake of completeness.

Reply:
Actually I'm removing that again. If we ever add refcounting in this area then we might want to make this more robust to exceptions thrown in between creating the response and returning it, and it'll be easier to consider those problems if we don't do any refcounting at all here rather than doing it badly.